### PR TITLE
CI: Add test case for unwanted patterns

### LINKED ISSUE #30454: CI: code check for " " introduced by black
Sometimes when we run `black` on something like

```
foo = (
    "bar "
    "baz"
)       
```

we end up with `foo = "bar " "baz"` when we would want `foo = "bar baz"`.

This is [not](https://github.com/psf/black/issues/1051) considered a bug in `black`, so is something we're currently checking for manually.  We should add it to code_checks.

### PR DESCRIPTION
**Pull request by ShaharNaveh**
- [x] closes #30454
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Can be merged, after merging #30464

### CODE REVIEW

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: You can simply use `if any(file_name.endswith(ext) for ext in [".pxd", ".py", ".pyx", ".pyx.ini"]):` and get rid of the `full_ext` function.

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: You could use something like `failures = 0` and later `failures += not is_concatenated(...`, you don't use the set of status codes.

But instead, I think it'd be better if `is_concatenated` is a generator and yields the errors.

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: I think it's more Pythonic (and easier to read) to use `for current_token, next_token in zip(tokens, tokens[1:]):`

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: I think something like `if current_token[TYPE] == next_token[TYPE] == token.STRING:` is much easier to understand.

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: I'd yield these values, and from the caller you'll now if something is broken or not depending if something is being yielded.

Probably worth renaming the function to something like "strings_to_concatenate" if this is what it's going to return.

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: I think is a good practice that this if gets the parameters from `sys.argv`, prints to stdout and returns the exit code. This way, the function that does the logic is well encapsulated with a clear input and output, and can be called from other parts of the code if needed.

**Review by datapythonista — CHANGES_REQUESTED** [review]
Nice work, added some comments about the structure of the script, and the style, but looks good.

**Comment by jbrockmendel** [comment]
This is really nice, thanks @MomIsBestFriend.  I was _not_ looking forward to figuring out the `ast` stuff.

@datapythonista is there someone on one of the flake/lint projects we might want to talk to about upstreaming this?

**Comment by jbrockmendel** [comment]
This is really nice, thanks @MomIsBestFriend.  I was _not_ looking forward to figuring out the `ast` stuff.

@datapythonista is there someone on one of the flake/lint projects we might want to talk to about upstreaming this?

**Comment by MomIsBestFriend** [comment]
@datapythonista It's looking like the running version of python in github's action is ```2.7.15```.

File: scripts/validate_string_concatenation.py
Comment (MomIsBestFriend) [reply scripts/validate_string_concatenation.py]: Next commit is going to use a generator.

**Comment by alimcmaster1** [comment]
> @datapythonista It's looking like the running version of python in github's action is `2.7.15`, is it possible to increase it?

Where does it say this out of interest? It must be at least 3.6 as we run `validate_docstring.py` in github actions?

File: scripts/validate_string_concatenation.py
Comment (alimcmaster1) [inline scripts/validate_string_concatenation.py]: Is this correct? Might be what's causing your issue mentioned https://github.com/pandas-dev/pandas/pull/30467#issuecomment-569092377

**Comment by MomIsBestFriend** [comment]
> > @datapythonista It's looking like the running version of python in github's action is `2.7.15`, is it possible to increase it?
> 
> Where does it say this out of interest? It must be at least 3.6 as we run `validate_docstring.py` in github actions?

https://github.com/pandas-dev/pandas/pull/30467/checks#step:4:52

File: scripts/validate_string_concatenation.py
Comment (MomIsBestFriend) [reply scripts/validate_string_concatenation.py]: > I'd yield these values, and from the caller you'll now if something is broken or not depending if something is being yielded.

Do you recommend yielding this values as list/tuple or as a string?

**Comment by datapythonista** [comment]
We never use the system python, we install conda in every build, and create an environment, you'll have to activate it.

I'd yield the values as a namedtuple.

**Comment by MomIsBestFriend** [comment]
> and create an environment, you'll have to activate it.


The problem is, the activation of conda is happening after it's running the unwanted patterns check. Do you have any suggestions?

**Comment by datapythonista** [comment]
Better move the string thing in the linting, which is a good idea anyway.

File: scripts/validate_string_concatenation.py
Comment (MomIsBestFriend) [inline scripts/validate_string_concatenation.py]: apeending a '3' solve the issue:)

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: Can you move `pxd` at the end please.

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: Why is this needed? `msg` should always evaluate to `True`

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: Same

**Review by datapythonista — COMMENTED** [review]
Looks great, added few comments with minor things, but good job. Thanks!

File: scripts/validate_string_concatenation.py
Comment (MomIsBestFriend) [inline scripts/validate_string_concatenation.py]: I'm really not sure about any of this

File: scripts/validate_string_concatenation.py
Comment (MomIsBestFriend) [inline scripts/validate_string_concatenation.py]: Unsure about the value of "azure"

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: make this a positional argument please, so instead of `script -p path` we simply call `script path`

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]:
```suggestion
        default="{source_path}:{line_number}:{start}:{end}:{msg}",
```
So by default we use this format, and if the caller ones a different format, can call the script like `script -f "##vso[task.logissue type=error; {source_path}]{msg}" path`. So you don't need to know the constants with the formats beforehand, just the variables.

I'm unsure whether `start` and `end` provide value, are they the columns of the strings to concatenate? I'd say the line number should make clear enough where the problem is, if that's the case.

Comment (MomIsBestFriend) [reply scripts/validate_string_concatenation.py]: > I'm unsure whether `start` and `end` provide value.

#### Short summary:

```start``` and ```end``` are used to pinpoint the unconcatenated string.

---

#### Long summary:

```start``` and ```end``` are the two unconcatenated strings that needs to be concatenated.

lets say we have a file called ```script.py``` and it's located in ```path/to/script.py``` and on line ```1337``` there is a line of:

```python
print("foo" "bar")
```

When running ```scripts/validate_string_concatenation.py```

The default the output is looking like this:

```
./path/to/script.py:1337      BETWEEN "foo" AND "bar"
```

and the yielded dictionary is mapped like this: 

```python
source_path: "./path/to/script.py"
line_number: 1337
start: "foo"
end: "bar"
```

File: ci/code_checks.sh
Comment (datapythonista) [inline ci/code_checks.sh]:
```suggestion
        $BASE_DIR/scripts/validate_string_concatenation.py . --format=""[error]{source_path}:{line_number}:\t BETWEEN {start} AND {end}"
```

This needs to have the format to be highlighted in GitHub actions

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: I think we can leave without this typing, doesn't add much value and decreases readability

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: Why do we need the `*`?

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]:
```suggestion
    dict of {str: str}
```

In the docstrings we don't use mypy syntax

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: Why do we need the `\n`? You're using print to display this, which already does the line break

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: The output of this will be something like:
```
pandas/__init__.py:124:    BETWEEN first part of the string AND with a second part in another string
```

I don't think people will understand what's going on when seeing this error in the CI. I think we should display something like:
```
pandas/__init__.py:124:String unnecessarily split in two by black. Please merge them manually.
```

File: ci/code_checks.sh
Comment (MomIsBestFriend) [reply ci/code_checks.sh]: Thank you :)

File: scripts/validate_string_concatenation.py
Comment (MomIsBestFriend) [reply scripts/validate_string_concatenation.py]: I can see what you mean. removed from code.

File: scripts/validate_string_concatenation.py
Comment (MomIsBestFriend) [reply scripts/validate_string_concatenation.py]: We don't, it's some leftover from when I built the script, just to organize my mind.

Removed from code.

File: scripts/validate_string_concatenation.py
Comment (MomIsBestFriend) [reply scripts/validate_string_concatenation.py]: I'm not sure about this, will the message be too long?
and where do I put the ```source_path, line_number, start, end```?

Comment (datapythonista) [reply scripts/validate_string_concatenation.py]: The message being long is not a problem. Neither for the CI or for the code.

From my example, the format would be:
```
{source_path}:{line_number}:String unnecessarily split in two by black. Please merge them manually.
```

As I mentioned eariler, I don't find `start` and `end` particularly useful for the reader of the error. Feel free to leave them if you disagree, but better rename them, their name is misleading. `string1` and `string2` are not great names, but they're better. At least they don't create the wrong impression those are the positions of the strings.

Comment (MomIsBestFriend) [reply scripts/validate_string_concatenation.py]: I can now see what you mean, Fixed.

Comment (datapythonista) [reply scripts/validate_string_concatenation.py]: Cool, looks much better.

Sorry I said it wrong, but the format should actually be `{source_path}:{line_number}:{msg}`. We'll decide the error here. Imagine that this function can detect more than one error, we can't say in the format which is the error we want to receive.

So this could become:
```suggestion
                    msg = "String unnecessarily split in two by black. Please merge them manually."
                    print(output_format.format(source_path=source_path, line_number=line_number, msg=msg))
```

And since now the `strings_to_concatenate` function is just returning two values, I think it makes more sense to return a tuple with them, instead of a dict, so the `for above` would be:
```
for source_path, line_number in strings_to_concatenate(os.path.join(subdir, file_name)):
```

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: Just list the two returns, instead of saying it's a tuple.

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: You don't need the parenthesis

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: Better just define `msg` once, since it's the same content for both cases.

**Review by datapythonista — COMMENTED** [review]
Cool, looks great now. Just couple of minor things, and I think we can get it merged. Thanks!

**Comment by MomIsBestFriend** [comment]
Can be merged after merging #30579

**Comment by datapythonista** [comment]
Merged #30579, you can update your branch, so the CI passes

File: scripts/validate_string_concatenation.py
Comment (datapythonista) [inline scripts/validate_string_concatenation.py]: the default is not true

**Review by datapythonista — COMMENTED** [review]
Just on small typo, but lgtm

**Review by datapythonista — APPROVED** [review]
lgtm, thanks @MomIsBestFriend

**Comment by MomIsBestFriend** [comment]
Thank you for the very close guidelines @datapythonista

**Comment by jreback** [comment]
thanks @MomIsBestFriend

**Pull request merged**
